Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift

```diff
@@ -0,0 +1,434 @@
+        // While there are a lot of Futures involved here it's quite possible that all of this code will
+        // actually complete synchronously: we just want to program for the possibility that it won't.
+        // Once that's done, we call the internal handler, then call the upgrader code, and then finally when the
+        // upgrader code is done, we do our final cleanup steps, namely we replay the received data we
+        // buffered in the meantime and then remove ourselves from the pipeline.
+        return {
+            // Ok, we're upgrading.
+            self.upgradeState = .upgrading
+            
+            self.removeHTTPHandlers(context: context)
+            .map {
+                // Let the other handlers be removed before continuing with upgrade.
+                self.upgradeCompletionHandler(context)
+                self.upgradeState = .upgradingAddingHandlers
+            }
+            .flatMap {
+                upgrader.upgrade(context: context, upgradeResponse: response)
+            }
+            .map {
+                self.upgradeState = .upgradeComplete
+                
+                // We unbuffer any buffered data here.
+                
+                // If we received any, we fire readComplete.
+                let fireReadComplete = self.receivedMessages.count > 0
+                while self.receivedMessages.count > 0 {
+                    let bufferedPart = self.receivedMessages.removeFirst()
+                    context.fireChannelRead(bufferedPart)
+                }
+                if fireReadComplete {
+                    context.fireChannelReadComplete()
+                }
+
+            }
+            .whenComplete { _ in
+                context.pipeline.removeHandler(context: context, promise: nil)
+            }
+        }
+    }
+    
+    /// Removes any extra HTTP-related handlers from the channel pipeline.
+    private func removeHTTPHandlers(context: ChannelHandlerContext) -> EventLoopFuture<Void> {
+        guard self.httpHandlers.count > 0 else {
+            return context.eventLoop.makeSucceededFuture(())
+        }
+        
+        let removeFutures = self.httpHandlers.map { context.pipeline.removeHandler($0) }
+        return .andAllSucceed(removeFutures, on: context.eventLoop)
+    }
+    
+    private func gotUpgrader(upgrader: @escaping (() -> Void)) {
+
+        guard case .awaitingUpgrader = self.upgradeState else {
+            preconditionFailure("Unexpected upgrader state: \(self.upgradeState)")
+        }
+
+        self.upgradeState = .upgraderReady(upgrader)
+        if self.seenFirstResponse {
+            // Ok, we're good to go, we can upgrade. Otherwise we're waiting for .end, which
+            // will trigger the upgrade.
+            upgrader()
+        }
+    }
+    
+    private func notUpgrading(context: ChannelHandlerContext, data: HTTPClientResponsePart) {
```

Comment:
Once we deal with the confusion around whether `handleUpgradeForProtocol` should return a `Future`, it should become clear in the code that `notUpgrading` is only ever synchronously called when we have received a `HTTPClientResponsePart.head`. With that in mind, the code here can be simplified, as we only ever need one branch: the first one.